### PR TITLE
Eliminate copying of paint_struct, initialize data in-place

### PR DESCRIPTION
**Pull request by ZehMatt**
Before:
sleepy_2021-01-07_10-13-44

After:
sleepy_2021-01-07_10-10-03

Small performance boost

### CODE REVIEW

**Comment by duncanspumpkin** [comment]
As discussed on discord. This undo's the structure i created recently as it was not as performant. It's a pity, but I'm not that attached to my old design. All looks fine to me. Would be good to do a good bit of testing on this to make sure there are no changes. Make sure peeps are holding the correct attachments, vehicles and entrances look correct. Make sure everything can be interacted with correctly.

**Comment by Gymnasiast** [comment]
Can you also post benchgfx results?

**Comment by ZehMatt** [comment]
> As discussed on discord. This undo's the structure i created recently as it was not as performant. It's a pity, but I'm not that attached to my old design. All looks fine to me. Would be good to do a good bit of testing on this to make sure there are no changes. Make sure peeps are holding the correct attachments, vehicles and entrances look correct. Make sure everything can be interacted with correctly.

Everything seems to be fine, can't see any visual glitches and interaction is still doing what it should.

> Can you also post benchgfx results?


```
develop:
Zoom[0] average: 0.080556s, 12 FPS
Zoom[1] average: 0.046600s, 21 FPS
Zoom[2] average: 0.029206s, 34 FPS
Total average: 0.052121s, 19 FPS
Time: 3.12725s

PR:
Zoom[0] average: 0.079380s, 13 FPS
Zoom[1] average: 0.045241s, 22 FPS
Zoom[2] average: 0.027784s, 36 FPS
Total average: 0.050802s, 20 FPS
Time: 3.04809s
```
Nothing too fancy here, software renderer has other bottlenecks so its not as prominent. With OpenGL I get between 5 and 10 FPS extra.

**Review by duncanspumpkin — APPROVED** [review]

File: src/openrct2/paint/Paint.cpp [hunk 1]
```diff
@@ -732,14 +732,13 @@ paint_struct* PaintAddImageAsParent(
     CoordsXYZ offset = { x_offset, y_offset, z_offset };
     CoordsXYZ boundBoxSize = { bound_box_length_x, bound_box_length_y, bound_box_length_z };
     CoordsXYZ boundBoxOffset = { bound_box_offset_x, bound_box_offset_y, bound_box_offset_z };
-    auto newPS = CreateNormalPaintStruct(session, image_id, offset, boundBoxSize, boundBoxOffset);
 
-    if (!newPS.has_value())
+    auto ps = CreateNormalPaintStruct(session, image_id, offset, boundBoxSize, boundBoxOffset);
```
Comment (Broxzier) [inline src/openrct2/paint/Paint.cpp:736]:
```suggestion
    auto* ps = CreateNormalPaintStruct(session, image_id, offset, boundBoxSize, boundBoxOffset);
```

File: src/openrct2/paint/Paint.cpp [hunk 2]
```diff
@@ -776,13 +775,14 @@ paint_struct* PaintAddImageAsParent(
     CoordsXYZ offset = { x_offset, y_offset, z_offset };
     CoordsXYZ boundBoxSize = { bound_box_length_x, bound_box_length_y, bound_box_length_z };
     CoordsXYZ boundBoxOffset = { bound_box_offset_x, bound_box_offset_y, bound_box_offset_z };
-    auto ps = CreateNormalPaintStruct(session, image_id, offset, boundBoxSize, boundBoxOffset);
 
-    if (!ps.has_value())
+    auto ps = CreateNormalPaintStruct(session, image_id, offset, boundBoxSize, boundBoxOffset);
```
Comment (Broxzier) [inline src/openrct2/paint/Paint.cpp:779]:
```suggestion
    auto* ps = CreateNormalPaintStruct(session, image_id, offset, boundBoxSize, boundBoxOffset);
```

File: src/openrct2/paint/Paint.cpp [hunk 3]
```diff
@@ -806,23 +806,22 @@ paint_struct* PaintAddImageAsChild(
...
     {
         return PaintAddImageAsParent(
             session, image_id, offset.x, offset.y, boundBoxLength.x, boundBoxLength.y, boundBoxLength.z, offset.z,
             boundBoxOffset.x, boundBoxOffset.y, boundBoxOffset.z);
     }
 
-    auto newPS = CreateNormalPaintStruct(session, image_id, offset, boundBoxLength, boundBoxOffset);
-
-    if (!newPS.has_value())
+    auto ps = CreateNormalPaintStruct(session, image_id, offset, boundBoxLength, boundBoxOffset);
```
Comment (Broxzier) [inline src/openrct2/paint/Paint.cpp:817]:
```suggestion
    auto* ps = CreateNormalPaintStruct(session, image_id, offset, boundBoxLength, boundBoxOffset);
```

**Comment by Gymnasiast** [comment]
@duncanspumpkin Did you test it?

**Comment by duncanspumpkin** [comment]
No I've not. I don't see any issues with it as it's basically reverting something i did before

**Comment by ZehMatt** [comment]
I did some basic play testing, everything behaves as expected.

**Pull request merged**